File: DSA Busted/Heap/IsBinaryTreeHeap.cpp

// https://practice.geeksforgeeks.org/problems/is-binary-tree-heap/1

class Solution {
  public:
    void countNodes(struct Node* tree, int &count){
        if(tree==NULL)
            return;
        
        countNodes(tree->left, count);
        countNodes(tree->right, count);
        
        count++;
        
    }
    bool isCBT(struct Node* root, int index, int &count){
        if(root==NULL)
            return true;
        
        if(index>=count)
            return false;
            
        bool left = isCBT(root->left, 2*index+1, count);
        bool right= isCBT(root->right, 2*index+2, count);
        
        return left && right;
    }
    bool isMaxHeap(struct Node* root){
     
        if(root->left == NULL && root->right==NULL)
            return true;
        else if(root->left==NULL && root->right!=NULL)
            return false;
        else if(root->left!=NULL && root->right==NULL){
            if(root->data>root->left->data)
                return true;
            else
                return false;
        }
        else{
            if(root->data>root->left->data && root->data>root->right->data){
                bool left= isHeap(root->left);
                bool right= isHeap(root->right);
                return left&&right;
            }
            else{
                return false;
            }
        }
    }
    bool isHeap(struct Node* tree) {
        // code here
        int index= 0;
        int count= 0;
        countNodes(tree, count);
        
        if(isCBT(tree, index, count) && isMaxHeap(tree)){
            return true;
        }
        else{
            return false;
        }
    }
};

File: DSA-Problems/Heap/IsBinaryTreeHeap.java

// https://practice.geeksforgeeks.org/problems/is-binary-tree-heap/1

class Solution {
    static class Node {
        int data;
        Node left, right;
        Node(int val) { this.data = val; }
    }

    private int countNodes(Node root) {
        if (root == null) return 0;
        return 1 + countNodes(root.left) + countNodes(root.right);
    }

    private boolean isCBT(Node root, int index, int nodeCount) {
        if (root == null) return true;
        if (index >= nodeCount) return false;

        return isCBT(root.left, 2 * index + 1, nodeCount) &&
               isCBT(root.right, 2 * index + 2, nodeCount);
    }

    private boolean isMaxHeap(Node root) {
        if (root.left == null && root.right == null) return true;

        if (root.right == null) return root.data >= root.left.data;

        return (root.data >= root.left.data && root.data >= root.right.data) &&
               isMaxHeap(root.left) && isMaxHeap(root.right);
    }

    public boolean isHeap(Node root) {
        int count = countNodes(root);
        return isCBT(root, 0, count) && isMaxHeap(root);
    }
}